Support config imports

mxpv opened the pull request:
Carry https://github.com/containerd/containerd/pull/3310
Fixes https://github.com/containerd/containerd/issues/3289

estesp commented:
Looks reasonable; just remembered we [have a man page](https://github.com/containerd/containerd/blob/master/docs/man/containerd-config.toml.5.md) for the config.toml file; it may already be missing some config options changes already (and definitely hasn't been updated for v2), but we should probably try and get that updated with info on how to use import and the rules for overwrite? Can be a separate PR

mxpv commented:
I've updated the merging logic a bit so imported files will append `slices` and `maps` instead of overwriting the whole section, and updated the config doc.

dmcgowan commented:
Merging maps makes sense to me, appending slices worries me a bit. Looking through the config I see a few places that couple be problematic. For example when configuring a list of endpoints, you may want to override the default but doing a slice append makes that impossible.

mxpv commented:
@dmcgowan  Slice overwrite can be more confusing IMO. For example if user wants to configure one specific plugin in a separate file, but instead overwrites the whole plugin section defined in parent config.

dmcgowan commented:
@mxpv I understand, getting correct can be very tricky. Generally if you want a configuration to be mergeable, using a map is sufficient. For example for plugins, the plugins have a name which allows this to be done pretty easily. The only case where I see this is potentially an issue is with the stream processors, but we still have a chance to make that a map for this release. Are there other parts you think this would be a problem? I think the current append method is workable since there is no way to force the overwrite.

crosbymichael commented:
What about just doing a replace based on the sections? I makes it easier to reason about and the user can just copy paste the section they want to change

mxpv commented:
Done. @dmcgowan @crosbymichael PTAL

fuweid commented on services/server/config/config.go:
```diff
@@ -198,23 +199,125 @@ func (c *Config) Decode(p *plugin.Registration) (interface{}, error) {
+		path, pending = pending[0], pending[1:]
+
+		// Check if a file at the given path already loaded to prevent circular imports
+		if _, ok := loaded[path]; ok {
```
nit: `if loaded[path]` more clear :P or use `map[string]struct{}`

dmcgowan commented:
@fuweid a command to dump the config would be good, maybe `containerd config dump`, since we already have `containerd config default`.

dmcgowan commented:
Merge logic looks good, I found one bug in the dump logic that am I seeing if I can fix myself. Also another bug in the proxy plugins but that is a separate issue I will add a PR for.

crosbymichael commented:
LGTM
